controllers/twophase/types.go

```diff
@@ -61,221 +61,80 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-			updateFailedMessage(ctx, req, r, chaos, err.Error())
-			return ctrl.Result{}, err
-		}
-
-		tempRecover := now.Add(*duration)
-		if tempStart.Before(tempRecover) {
-			err := fmt.Errorf("nextRecover shouldn't be later than nextStart")
-			r.Log.Error(err, "Then recover can never be reached.", "scheduler", *chaos.GetScheduler(), "duration", *duration)
-			updateFailedMessage(ctx, req, r, chaos, err.Error())
-			return ctrl.Result{}, err
-		}
-
-		if err = applyAction(ctx, r, req, *duration, chaos); err != nil {
-			updateFailedMessage(ctx, req, r, chaos, err.Error())
-			return ctrl.Result{Requeue: true}, err
-		}
-
-		nextStart, err := utils.NextTime(*chaos.GetScheduler(), status.Experiment.StartTime.Time)
-		if err != nil {
-			r.Log.Error(err, "failed to get the next start time")
-			return ctrl.Result{}, err
-		}
-		nextRecover := status.Experiment.StartTime.Time.Add(*duration)
-
-		chaos.SetNextStart(*nextStart)
-		chaos.SetNextRecover(nextRecover)
-		status.FailedMessage = emptyString
-	} else {
-		r.Log.Info("Waiting")
-
-		nextStart, err := utils.NextTime(*chaos.GetScheduler(), status.Experiment.StartTime.Time)
-		if err != nil {
-			r.Log.Error(err, "failed to get next start time")
-			return ctrl.Result{}, err
-		}
-		nextTime := chaos.GetNextStart()
-
-		// if nextStart is not equal to nextTime, the scheduler may have been modified.
-		// So set nextStart to time.Now.
-		if nextStart.Equal(nextTime) {
-			if !chaos.GetNextRecover().IsZero() && chaos.GetNextRecover().Before(nextTime) {
-				nextTime = chaos.GetNextRecover()
-			}
-			duration := nextTime.Sub(now)
-			r.Log.Info("Requeue request", "after", duration)
+		targetPhase = v1alpha1.ExperimentPhaseFinished
+	}
 
-			return ctrl.Result{RequeueAfter: duration}, nil
-		}
+	r.Log.Info("decide target phase", "target phase", targetPhase)
 
-		chaos.SetNextStart(time.Now())
-		duration := nextTime.Sub(now)
-		r.Log.Info("Requeue request", "after", duration)
-		return ctrl.Result{RequeueAfter: duration}, nil
+	machine := chaosStateMachine{
+		Chaos:      chaos,
+		Req:        req,
+		Reconciler: r,
 	}
-
-	if err := r.Update(ctx, chaos); err != nil {
-		r.Log.Error(err, "unable to update chaos status")
+	err = machine.Into(ctx, targetPhase, now)
+	if err != nil {
+		r.Log.Error(err, "fail to step into the phase", "target phase", targetPhase)
 		return ctrl.Result{}, err
 	}
 
-	return ctrl.Result{}, nil
-}
-
-func applyAction(
-	ctx context.Context,
-	r *Reconciler,
-	req ctrl.Request,
-	duration time.Duration,
-	chaos v1alpha1.InnerSchedulerObject,
-) error {
-	status := chaos.GetStatus()
-	r.Log.Info("Chaos action:", "chaos", chaos)
-
-	// Start to apply action
-	r.Log.Info("Performing Action")
-
-	if err := r.Apply(ctx, req, chaos); err != nil {
-		r.Log.Error(err, "failed to apply chaos action")
-
-		status.Experiment.Phase = v1alpha1.ExperimentPhaseFailed
+	if chaos.GetStatus().Experiment.Phase == v1alpha1.ExperimentPhaseFinished {
```

Comment:
In the previous commit version, if it's from Paused to Paused, our controller would only run into [Line 114](https://github.com/chaos-mesh/chaos-mesh/pull/1221/files#diff-04c792dfed57630c1de983400da2e7639c6d5c3a929c0124f7b5c87aa9a19105L114), and don't re-enqueue it.

To fix the `Paused->Paused` case, how about make the changes like below:

```suggestion
	if chaos.GetStatus().Experiment.Phase == v1alpha1.ExperimentPhaseFinished || chaos.GetStatus().Experiment.Phase == v1alpha1.ExperimentPhasePaused {
```

Reply:
Good Idea :+1: